Sources/SpeziHealthKit/HealthKit.swift

```diff
@@ -68,129 +69,215 @@ import SwiftUI
+        switch healthStore.authorizationStatus(for: objectType) {
+        case .notDetermined:
+            false
+        case .sharingAuthorized, .sharingDenied:
+            true
+        @unknown default:
+            false // fallback. better be safe than sorry.
+        }
+    }
+    
+    /// Returns whether the application is currently authorized to write data of the specified object type into the health store..
+    /// - Note: A `false` return value does **not** imply that the user actually denied access; it could also mean that the user hasn't yet been asked.
+    ///     Use ``HealthKit-swift.class/askedForAuthorization(toWrite:)`` to determine that.
+    public func isAuthorized(toWrite objectType: HKObjectType) -> Bool {
+        switch healthStore.authorizationStatus(for: objectType) {
+        case .sharingAuthorized:
+            true
+        case .notDetermined, .sharingDenied:
+            false
+        @unknown default:
+            false // fallback. better be safe than sorry.
+        }
+    }
+    
+    
+    // MARK: HealthKit data collection
+    
+    /// Adds a new data source for collecting health data in the background.
     @MainActor
-    public func execute(@HealthKitDataSourceDescriptionBuilder _ healthKitDataSourceDescriptions: () -> [HealthKitDataSourceDescription]) {
-        for healthKitDataSourceDescription in healthKitDataSourceDescriptions() {
-            execute(healthKitDataSourceDescription)
+    func addBackgroundHealthDataSource(_ dataSource: some HealthKitDataSource) async {
+        registeredDataSources.append(dataSource)
+        if await askedForAuthorization(toRead: dataSource.sampleType) {
+            // If we already asked for authentication to this specific sample type, we can directly start it.
+            // Otherwise, the next call to askForAuthorization will trigger the start of the data collection.
+            Task {
+                await dataSource.startAutomaticDataCollection()
+            }
         }
     }
     
-    /// Triggers any ``HealthKitDeliverySetting/manual(safeAnchor:)`` collections and starts the collection for all ``HealthKitDeliveryStartSetting/manual`` HealthKit data collections.
+    /// Triggers any ``HealthKitDeliverySetting/manual(saveAnchor:)`` collections and starts the collection for all ``HealthKitDeliveryStartSetting/manual`` HealthKit data collections.
     @MainActor
     public func triggerDataSourceCollection() async {
         await withTaskGroup(of: Void.self) { group in
-            for healthKitComponent in healthKitComponents {
+            for dataSource in registeredDataSources {
                 group.addTask { @MainActor @Sendable in
-                    await healthKitComponent.triggerManualDataSourceCollection()
+                    await dataSource.triggerManualDataSourceCollection()
                 }
             }
             await group.waitForAll()
         }
     }
 }
+
+
+
+
+
+extension HKHealthStore {
+    // TODO remove?!
```

Comment:
What would be the reason for this function?

Reply:
It'll get removed, but there are some object types where you need to request per-object access (instead of a simple one-time request covering all objects of the type).
we're currently not using this; i was thinking about maybe adding support but i then concluded that it'd exceed the scope of the PR and probably (right now, at least) also of the HealthKit package as a whole.

i had some issues calling the async version of that function and added a custom overload to work around that.